timer, cache, retry decorators implementation with custom tests

konpsar opened the pull request:
Implemented 3 python decorators: timer, cache, retry with some arguments each. Implemented also a few tests running utilizing each decorator with/without arguments (no use of any testing library or any asserts. Just manual testing with some output observation).

konpsar committed:
Create decorators.py

KonstantinosPsaroulakis committed:
add timer decorator w/ some tests

KonstantinosPsaroulakis committed:
added timer_v2 with output file support

KonstantinosPsaroulakis committed:
implemented cache deco + split decorators testing to sep files

KonstantinosPsaroulakis committed:
cache decorator:: added support for cache size++debug msgs

KonstantinosPsaroulakis committed:
added retry decorator implementation + simple tests

KonstantinosPsaroulakis committed:
retry decorator:: added delay support

KonstantinosPsaroulakis committed:
retry decorator:: added backoff support

KonstantinosPsaroulakis committed:
change definitions to contain dtypes as well

KonstantinosPsaroulakis committed:
ignore logs output

vbardakos commented on python/decorators.py:
```diff
@@ -0,0 +1,59 @@
+    cache = {}
+    def decorator(func):
+        def wrapper(*args, **kwargs):
+            key = args
```
The current version doesn't take `kwargs` into account, which means:
```python
@cache_decorator(5)
def foo(x: float, y: float) -> float:
    return x + y

foo(1, y=1)  # not cached --> result: 2
foo(1, y=9)  # cached     --> result: 2
```

Note:
There's no reason for now to try to cache all possible parameter combinations correctly:
```
(1, 2) == (1, y=2) == (x=1, y=2)
```

vbardakos commented on python/decorators.py:
```diff
@@ -0,0 +1,59 @@
+            key = args
+            print(f"---> Current cache (start): {cache}")
+            print(f'------> Cached value for args: {key}: {cache.get(key, 'Not cached')}')
+            if key not in cache:
+                if size and len(cache) >= size:
+                    cache.popitem()
+                cache[key] = func(*args, **kwargs)
+            print(f"---------> Current cache (end): {cache}")
```
A `dict` is definitively the most elegant way to approach this problem. However, this might not be the most performant data structure in this use case. Can you think of an alternative which would make sense?

vbardakos commented on python/decorators.py:
```diff
@@ -0,0 +1,59 @@
+                    actual_delay = delay * max(backoff * t, 1)
+                    print(f"Retrying in {actual_delay} seconds")
+                    time.sleep(actual_delay)
+            return None
```
In the current implementation, the decorator will handle every attempt and return None. Although, this prevents explicit error handling in case all of the retries fail. 

Conceptually, a retry decorator is a utility decorator on functions where you try to "eliminate" some kind of uncertainty (eg a simple timeout vs database being down) before you explicitly handle the issue.

vbardakos commented on python/decorators.py:
```diff
@@ -0,0 +1,59 @@
+    return decorator
+
+def cache_decorator(size: int = None):
+    cache = {}
```
What if I want clear the cache during runtime? How can we add this feature to the decorator?

vbardakos commented on python/decorators.py:
```diff
@@ -0,0 +1,59 @@
+            for t in range(tries):
+                try:
+                    return func(*args, **kwargs)
+                except Exception as e:
```
Why to handle `Exception` and not `BaseException`?

vbardakos commented on python/decorators.py:
```diff
@@ -0,0 +1,59 @@
+        end_time = time.time()
+        execution_time = end_time - start_time
+        print(f"Execution time: {execution_time} seconds")
+        return result
```
Is there a way to make our decorator more robust?

KonstantinosPsaroulakis committed:
cache decorator :: add support for kwargs

KonstantinosPsaroulakis committed:
make timer decorator more generic to get a Writable as argument + can be used without ()

KonstantinosPsaroulakis committed:
replace time.time with time.perfcounter -- highest resolution timer --

KonstantinosPsaroulakis committed:
Modified timer decorator to work regardless of any exceptions raised

KonstantinosPsaroulakis committed:
removed '_decorator' from all decorators' names

KonstantinosPsaroulakis committed:
fix retry decorator so in the last try, no error handling is performed

KonstantinosPsaroulakis committed:
Add clear_cache() function to cache decorator for manual cache clearing during runtime

KonstantinosPsaroulakis committed:
Re-implement cache decorator with deque instead of dict

An orderedDict may be a better choice though. To keep the key-value relations along with the ordering to more efficient caching using a FIFO concept

KonstantinosPsaroulakis committed:
add exception handling to timer in case of non writable output + implement pytests

KonstantinosPsaroulakis committed:
implement retry with the guard clause at the start

KonstantinosPsaroulakis committed:
implement deque cache with maxlen + corresponding usage examples (not unit tests)

KonstantinosPsaroulakis committed:
use pytest mocker in timer testing ++ add given/when/then comments

KonstantinosPsaroulakis committed:
move dummy_function out of tests to reuse it
